From 90132d7394bba59e95b37f1cb6b0533a4f92a064 Mon Sep 17 00:00:00 2001
From: Shikha Vyaghra <vyaghras@amazon.com>
Date: Thu, 6 Jul 2023 17:26:45 +0000
Subject: [PATCH] Change default capabilities using daemon config

Default capabilities in spec can be changed by reading from daemon
configuration file using a parameter "default-capabilities". If
the capabilities will not be provided, then default capabilities
in Moby code will be used.

Signed-off-by: Shikha Vyaghra <vyaghras@amazon.com>
[agarrcia: updated for docker v25]
Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>
[updated for docker v29]
Signed-off-by: Kush Upadhyay <kushupad@amazon.com>
[Rebased for Docker 29.9.0; preserve nri-opts and extension-config.]
---
 daemon/command/config_unix.go |  1 +
 daemon/config/config.go       |  6 ++++--
 daemon/config/config_linux.go |  5 +++--
 daemon/oci_linux.go           | 13 ++++++++++---
 4 files changed, 18 insertions(+), 7 deletions(-)

diff --git a/daemon/command/config_unix.go b/daemon/command/config_unix.go
index c086dd5..880270b 100644
--- a/daemon/command/config_unix.go
+++ b/daemon/command/config_unix.go
@@ -58,4 +58,5 @@ func installConfigFlags(conf *config.Config, flags *pflag.FlagSet) {
 	// Note that conf.BridgeConfig.UserlandProxyPath and honorXDG are configured according to the value of rootless.RunningWithRootlessKit, not the value of --rootless.
 	flags.BoolVar(&conf.Rootless, "rootless", conf.Rootless, "Enable rootless mode; typically used with RootlessKit")
 	flags.StringVar(&conf.CgroupNamespaceMode, "default-cgroupns-mode", conf.CgroupNamespaceMode, `Default mode for containers cgroup namespace ("host" | "private")`)
+	flags.Var(opts.NewNamedListOptsRef("default-capabilities", &conf.Capabilities, nil), "default-capabilities", "Default capabilities for containers")
 }
diff --git a/daemon/config/config.go b/daemon/config/config.go
index e03b04e..a40eeca 100644
--- a/daemon/config/config.go
+++ b/daemon/config/config.go
@@ -91,6 +91,7 @@
 	"default-ulimits":      true,
 	"features":             true,
 	"builder":              true,
+	"default-capabilities": true,
 	"nri-opts":             true,
 	"extension-config":     true,
 }
@@ -99,8 +100,9 @@
 // that will be skipped from findConfigurationConflicts
 // for unknown flag validation.
 var skipValidateOptions = map[string]bool{
-	"features": true,
-	"builder":  true,
+	"features":             true,
+	"builder":              true,
+	"default-capabilities": true,
 
 	// Only available in daemon.json, no flags
 	"min-api-version":  true,
diff --git a/daemon/config/config_linux.go b/daemon/config/config_linux.go
index 1d8138a..3be9206 100644
--- a/daemon/config/config_linux.go
+++ b/daemon/config/config_linux.go
@@ -95,8 +95,9 @@ type Config struct {
 	IpcMode              string                       `json:"default-ipc-mode,omitempty"`
 	CgroupNamespaceMode  string                       `json:"default-cgroupns-mode,omitempty"`
 	// ResolvConf is the path to the configuration of the host resolver
-	ResolvConf string `json:"resolv-conf,omitempty"`
-	Rootless   bool   `json:"rootless,omitempty"`
+	ResolvConf   string   `json:"resolv-conf,omitempty"`
+	Rootless     bool     `json:"rootless,omitempty"`
+	Capabilities []string `json:"default-capabilities,omitempty"`
 }
 
 // GetExecRoot returns the user configured Exec-root
diff --git a/daemon/oci_linux.go b/daemon/oci_linux.go
index 1b3bbc5..914cd9e 100644
--- a/daemon/oci_linux.go
+++ b/daemon/oci_linux.go
@@ -156,10 +156,17 @@ func WithApparmor(daemon *Daemon, c *container.Container) coci.SpecOpts {
 
 // WithCapabilities adjusts the container's capabilities based on the
 // "CapAdd", "CapDrop", and "Privileged" fields in the container's HostConfig.
-func WithCapabilities(ctr *container.Container) coci.SpecOpts {
+func WithCapabilities(ctr *container.Container, daemonCfg *dconfig.Config) coci.SpecOpts {
 	return func(ctx context.Context, client coci.Client, c *containers.Container, s *specs.Spec) (err error) {
+		var defCaps []string
+		if len(daemonCfg.Capabilities) != 0 {
+			defCaps = daemonCfg.Capabilities
+		} else {
+			defCaps = caps.DefaultCapabilities()
+		}
+
 		capabilities, err := caps.TweakCapabilities(
-			caps.DefaultCapabilities(),
+			defCaps,
 			ctr.HostConfig.CapAdd,
 			ctr.HostConfig.CapDrop,
 			ctr.HostConfig.Privileged,
@@ -1032,7 +1039,7 @@ func (daemon *Daemon) createSpec(ctx context.Context, daemonCfg *configStore, c
 		WithDevices(daemon, c),
 		withRlimits(daemon, &daemonCfg.Config, c),
 		WithNamespaces(daemon, c),
-		WithCapabilities(c),
+		WithCapabilities(c, &daemonCfg.Config),
 		WithSeccomp(daemon, c),
 		withMounts(daemon, daemonCfg, c, mounts),
 		WithApparmor(daemon, c),
